### PR TITLE
Udaf n gram

### PR DESCRIPTION
**Pull request by gcz2022**
## What changes were proposed in this pull request?

This is the initial version of UDAF NGram.
Multiple issues will be taken care of next:
* Change to `java.util.HashMap`                                                                      **Done**
* Support for `Array of Array of String`  **Done**
* The use of `accuracy` and `HashMap`'s trim while growing too big              **Done**
* The last returning result format **Done**
* Manual serialization to gain better perf
* Choice of Array/List to get better perf  **Done**
* More test cases in Spark SQL core  **Done**
* NULL support 
......

## How was this patch tested?

Mainly refer to `NGramsQuerySuite` in spark sql core module , due to `NGramsSuite` will be removed in the future.

### CODE REVIEW

**Comment by gcz2022** [comment]
cc @chenghao-intel @adrian-wang @yucai
Could you please help me review this alpha version code(maybe in general view)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    }
+  }
+
+  override def createAggregationBuffer(): NGramBuffer = {
+    new NGramBuffer(n, k, accuracy, new OpenHashMap[List[UTF8String], Double]())
+  }
+
+  override def update(buffer: NGramBuffer, inputRow: InternalRow): NGramBuffer = {
+    val genericArrayData: GenericArrayData = child.eval(inputRow).asInstanceOf[GenericArrayData]
+    val values = (0 until genericArrayData.numElements()).map(genericArrayData.get(_, StringType).asInstanceOf[UTF8String]).toList
```
Comment (yucai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: Why not use genericArrayData.numElements directly?

Comment (gcz2022) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: Where to use it directly? The `values` represents a list of string.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      TypeCheckFailure(
+        s"The accuracy provided must be a positive integer literal (current value = $accuracy)")
+    }
+    else {
+      TypeCheckSuccess
+    }
+  }
+
+  override def createAggregationBuffer(): NGramBuffer = {
+    new NGramBuffer(n, k, accuracy, new OpenHashMap[List[UTF8String], Double]())
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: using the `java.util.HashMap` instead.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  override def update(buffer: NGramBuffer, inputRow: InternalRow): NGramBuffer = {
+    val genericArrayData: GenericArrayData = child.eval(inputRow).asInstanceOf[GenericArrayData]
+    val values = (0 until genericArrayData.numElements()).map(genericArrayData.get(_, StringType).asInstanceOf[UTF8String]).toList
+    if (values != null) {
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: I don't think `values` can be `null`, probably we don't need this check.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      val nGrams = getNGrams(values, n)
+      nGrams.foreach(buffer.add(_))
+    }
+    buffer
+  }
+
+  private def getNGrams(values: List[UTF8String], n: Int): List[List[UTF8String]] = {
+    values.sliding(n).toList
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: Why do you need `toList`? will it cause extra performance penalty?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      TypeCheckSuccess
+    }
+  }
+
+  override def createAggregationBuffer(): NGramBuffer = {
+    new NGramBuffer(n, k, accuracy, new OpenHashMap[List[UTF8String], Double]())
+  }
+
+  override def update(buffer: NGramBuffer, inputRow: InternalRow): NGramBuffer = {
+    val genericArrayData: GenericArrayData = child.eval(inputRow).asInstanceOf[GenericArrayData]
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: `genericArrayData` probably be null.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 6]
```diff
@@ -0,0 +1,175 @@
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import java.nio.ByteBuffer
+
+import com.google.common.primitives.{Chars, Doubles, Ints}
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: add empty line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{TypeCheckFailure, TypeCheckSuccess}
+import org.apache.spark.sql.catalyst.expressions.aggregate.NGrams.NGramBuffer
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription, ImplicitCastInputTypes, Literal}
+import org.apache.spark.sql.catalyst.util.{GenericArrayData, QuantileSummaries}
+import org.apache.spark.sql.types._
+import org.apache.spark.util.collection.OpenHashMap
+import org.apache.spark.serializer.KryoSerializer
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: import order..

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGramsSuite.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.catalyst.expressions.{BoundReference, Cast, CreateArray, ExpressionEvalHelper, Literal}
+import org.apache.spark.sql.catalyst.util.GenericArrayData
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+  * Created by gcz on 17-2-20.
```
Comment (chenghao-intel) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGramsSuite.scala]: We don't leave the author information in the code in Spark project.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGramsSuite.scala [hunk 9]
```diff
@@ -0,0 +1,103 @@
+package org.apache.spark.sql.catalyst.expressions.aggregate
```
Comment (chenghao-intel) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGramsSuite.scala]: Add Licence.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGramsSuite.scala [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        boolean ${ev.isNull} = ${ctx.INPUT_ROW}.isNullAt($size);
+        $javaType ${ev.value} = ${ev.isNull} ? ${ctx.defaultValue(dataType)} : ($value);""")
+    } else {
+      ev.copy(code = s"""$javaType ${ev.value} = $value;""", isNull = "false")
+    }
+  }
+}
+
+
+class NGramsSuite extends SparkFunSuite with ExpressionEvalHelper{
```
Comment (chenghao-intel) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGramsSuite.scala]: space before `{`

**Comment by chenghao-intel** [comment]
You also need to register the nGram in `FunctionRegistry`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala (hunk 7 above)
Comment (gcz2022) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: Actually, these issues weren't taken care of due to this is initial version, I will deal with the major issues(as indicated in the front) first.
Thanks for your review, good night : )

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+import java.nio.ByteBuffer
+
+import com.google.common.primitives.{Chars, Doubles, Ints}
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{TypeCheckFailure, TypeCheckSuccess}
+import org.apache.spark.sql.catalyst.expressions.aggregate.NGrams.NGramBuffer
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription, ImplicitCastInputTypes, Literal}
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: import order

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 12]
```diff
@@ -0,0 +1,175 @@
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import java.nio.ByteBuffer
+
+import com.google.common.primitives.{Chars, Doubles, Ints}
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.{TypeCheckFailure, TypeCheckSuccess}
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: better use _

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala (hunk 11 above)
Comment (adrian-wang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: line 11 shoule be placed before line 10

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 13]
```diff
@@ -0,0 +1,175 @@
+package org.apache.spark.sql.catalyst.expressions.aggregate
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala:18]: license

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala (hunk 11 above)
Comment (gcz2022) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: Sorry, why? aggregate `a` is before Expression `e`

**Comment by gcz2022** [comment]
This PR is personally ready except for the manual serde and Null support.
For Null support, Hive will raise an Execution error caused by `java.lang.NullPointerException`, which I think is not thoughtful. I am not certain here, should I just skip the null to calculate top k ngrams or just report an error when null is found?
Could you please help me review it?
@yucai @adrian-wang @chenghao-intel

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+import java.nio.ByteBuffer
+import java.util.HashMap
+
+import org.apache.spark.serializer.KryoSerializer
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult._
+import org.apache.spark.sql.catalyst.expressions.aggregate.NGrams.NGramBuffer
+import org.apache.spark.sql.catalyst.expressions.{Expression, ExpressionDescription, ImplicitCastInputTypes, Literal}
```
Comment (gcz2022) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: I get a `0: import.ordering.wrongOrderInGroup.message` in line 28, does anyone know why?

Comment (adrian-wang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: line 28 should be placed before line 27

Comment (gcz2022) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala]: Thanks, what's the rule here ? `{` or group import should be placed in the former?

**Comment by gcz2022** [comment]
How to decline user's ngrams query with null? If I check children's type in `checkInputTypes` to check if `child.dataType == ArrayType(StringType, true)` will not succeed, because the equation will be true even if there is no null in the array.

**Pull request closed without merging**